src/pubsub.c

```diff
@@ -763,9 +763,21 @@ void channelList(client *c, sds pat, dict **pubsub_channels, int is_sharded) {
 
 /* SPUBLISH <shardchannel> <message> */
 void spublishCommand(client *c) {
-    int receivers = pubsubPublishMessageAndPropagateToCluster(c->argv[1],c->argv[2],1);
-    if (!server.cluster_enabled)
+    int receivers = pubsubPublishMessage(c->argv[1], c->argv[2], 1);
+    if (server.cluster_enabled) {
+        if (iAmMaster()) {
+            /* Propagate the message to other replica(s) in the shard via replication link. */
+            forceCommandPropagation(c, PROPAGATE_REPL);
+        } else if (!(c->flags & CLIENT_MASTER)) {
+            /* If the command is received from a regular client, propagate it via cluster link. */
+            clusterPropagatePublish(c->argv[1], c->argv[2], 1);
```

Comment:
I suppose this is an orthogonal concern, but it's hard for clients to stay attached to the master in the case of failovers. When a failover occurs, the client will continue to send messages to the replica and they will be inefficiently routed along the clusterbus. I suppose my preference is we also introduce a flag to force it be redirected to the primary in some way.

Reply:
I've tried to introduce a flag to force this behavior. PTAL.